### PR TITLE
Bug 1888073: prevent no-op hotlooping on Operators

### PR DESCRIPTION
**Pull request by sjenning**
xref https://bugzilla.redhat.com/show_bug.cgi?id=1888073

The `Operator` Reconciler does not detect when a reconciliation is triggered by its own Update and thus hotloops until it gets lucky enough to create the exact same rendering of the `Operator` two times in a row.  For larger `Operator` resources, this is a unlikely and will hotloop forever.

This hotloop creates a lot of strain on the kube-apiserver and etcd as well, since some `Operator` resources are very large.  I observe about 500m of additional CPU cores usage per master in addition to 1 core of CPU used directly by `olm-operator` hotlooping. `olm-operator` also generates 1-2MB/s of network traffic in this state.

Before patch with ACM Operator installed

olm-operator

After patch.  The usage spike is during ACM installation, but it falls back down after installation is compelete, as expected.

olm-operator-patched

@njhale @derekwaynecarr @dinhxuanvu

### CODE REVIEW

**Comment by sjenning** [comment]
/hold

Still some things to work out. It isn't recreating the Operator when it is deleted manually.  And the unit test failure I think is due to not requeuing on failure.

File: pkg/controller/operators/operator_controller.go
Comment (njhale) [inline pkg/controller/operators/operator_controller.go:154]:
```suggestion
		if err := r.Create(context.Background(), operator.Operator); err != nil && !apierrors.IsAlreadyExists(err) {
```

File: pkg/controller/operators/operator_controller.go
Comment (njhale) [inline pkg/controller/operators/operator_controller.go:155]: Is there a reason we don't want to return an error here and requeue? Any error at this point should be transient, and the work queue should rate limit to prevent hot loops.

File: pkg/controller/operators/operator_controller.go
Comment (njhale) [inline pkg/controller/operators/operator_controller.go:164]: are we sure that the Create and Status.Update() methods bump the ResourceVersion on the resource argument?

Comment (njhale) [inline pkg/controller/operators/operator_controller.go:164]:
```suggestion
	r.setLastResourceVersion(req.NamespacedName, operator.GetResourceVersion())
```

File: pkg/controller/operators/operator_controller.go
Comment (njhale) [reply pkg/controller/operators/operator_controller.go:154]: Note: this change makes the patch pass e2e and unit tests

File: pkg/controller/operators/operator_controller.go
Comment (sjenning) [reply pkg/controller/operators/operator_controller.go:165]: Yes, I checked this.  That is why Create/Update take pointers to the resource so they can return the resource after the server has processed it.

File: pkg/controller/operators/operator_controller.go
Comment (sjenning) [reply pkg/controller/operators/operator_controller.go:155]: Yes, we do want to requeue.  Updated.

**Comment by sjenning** [comment]
`Operator` is now recreating on delete and I think the unit test should be fixed now as well
/hold cancel

File: pkg/controller/operators/operator_controller.go
Comment (njhale) [inline pkg/controller/operators/operator_controller.go:144]: I don't think an error like this is transient; more likely it would be a programming error. If that's the case, wouldn't requeuing just congest the queue with something that would never reconcile correctly?

related: if we agree that an error here would mean the program is incorrect, then is a panic more appropriate?

**Review by njhale — COMMENTED** [review]
Wow, awesome work! Thanks again for finding this and patching it so quickly!

I have one last comment before I sign off. Let me know what you think.

**Comment by njhale** [comment]
looks like there's some flakiness in the unit tests

/hold

**Comment by sjenning** [comment]
`e2e-aws-olm` seems to have a <50% pass rate over all PRs

/test e2e-aws-olm

**Comment by sjenning** [comment]
The `End-to-end: CSV update with owned API service` e2e test is very flaky
/retest

**Comment by sjenning** [comment]
clean run :tada:

File: pkg/controller/operators/operator_controller.go
Comment (ecordell) [reply pkg/controller/operators/operator_controller.go:144]: this function (currently) can only error out when `in` is nil, which will never be the case here.

**Review by ecordell — APPROVED** [review]
/lgtm

**Pull request merged**
